nixos/modules/misc/documentation.nix

```diff
@@ -294,16 +294,36 @@ in
         '';
       };
 
-      nixos.extraModuleSources = mkOption {
-        type = types.listOf (types.either types.path types.str);
+      nixos.extraModuleSources = let
+        moduleSourceType = types.submodule {
+          options = {
+            prefix = mkOption {
+              type = types.path;
+              description = lib.mdDoc "The prefix under which modules from this source are found.";
+            };
+            name = mkOption {
+              type = with types; nullOr (strMatching "[a-zA-Z0-9._+-]+");
+              default = null;
```

Comment:
The problem is compatibility: previously `extraModuleSources` only accepted paths to be stripped, so the question is what to do with those. Currently they are mapped to a nameless source, I don't see a better alternative short of throwing an error, but that seems a bit brutal.

Reply:
`<_unknown/*>` may be a good default?

Reply:
Perhaps `baseNameOf config.prefix` and maybe strip off a store path hash part, if there is one.
I guess that will effectively be `"source"` in many cases.

What do the angle brackets, underscore, slash and asterisk mean? It seems a bit excessive.

Reply:
@roberth Imagine a NixOS module source under version control, checked out in a home directory. If there's a module at the top level, e.g. `$CHECKOUT/configuration.nix`, the base name of `$CHECKOUT` isn't necessarily stable between checkouts. We should avoid looking at parents of the top-level. If we don't provide a default module source name, this path renders in documentation as `<configuration.nix>`. If the source then imports a module from `$CHECKOUT/nixpkgs/nixos/system/build.nix`, its path then renders as `<nixpkgs/nixos/system/build.nix>`, which is the path of an existing Nixpkgs module. (This example is contrived, but the situation can reasonably occur if multiple external module sets use simple paths.) On the other hand, if we provide a default module source name of `_unknown/`, then the former path renders as `<_unknown/configuration.nix>` and the latter path renders as `<_unknown/nixpkgs/nixos/system/build.nix>`.

This approach can still result in collisions, but a default name of `_unknown` sticks out and looks ugly enough that anyone who notices will change it, and the leading underscore is conventional for internal Nix & Nixpkgs names.

If we wanted the name to stick out more and be easier to search, `_unknown-module-source` could work really well (`<_unknown-module-source/configuration.nix>`).

On the `…/configuration.nix` note from earlier, we may want to add `config.nixos.extraModuleSources = [ { name = "nixos-config"; prefix = ./.; } ];` to the `/etc/nixos/configuration.nix` template.

Reply:
Ok, now I understand what you meant.

I'd be quite happy for it to throw an exception in projects where I control the scope of docs generation actually, but for NixOS, returning something ugly seems like a pretty good, robust solution.

We can't really guess the name, and we certainly can't let the NixOS evaluation fail because of the docs, so `<_unknown/whatever/path/file.nix>` seems like a good solution.

Reply:
Agreed, will do this.

> we may want to add `config.nixos.extraModuleSources = [ { name = "nixos-config"; prefix = ./.; } ];` to the `/etc/nixos/configuration.nix` template.

I'm not convinced. Seems like a lot of boilerplate for something the vast majority of people won't care about (and if that file is indeed destined to be in `/etc/nixos/configuration.nix`, then `<nixos-config/X>` isn't a lot better than `/etc/nixos/X`.

Reply:
Actually, here's an alternative: introduce a new `moduleSources` option with the "ideal" type (attrsOf {prefix, urlFor}), and add a `mkChangedOptionModule` that maps paths from `extraModuleSources` to attributes in `moduleSources` using the mechanism proposed by @roberth (basename of prefix minus store hash). Seems like the best of all worlds?

Reply:
I guess this will break if someone has two paths named `source`, which is quite likely...

Reply:
I thought `name` was only a label for presentation, not a lookup key. Isn't prefix the only lookup key?

Reply:
Lookup key in the logical sense. I think prefix has to be a linear search, not an attrset key, because it can be a path value, and we may not want to `toString` those because of lazy trees.

Reply:
See the mirror [discussion](https://github.com/NixOS/nixpkgs/pull/199363#discussion_r1021733671) for `searchPath`. Technically you're right, but I think it makes sense to have module sources indexed by their name. We might find more use cases for them later than just stripping prefixes.

Reply:
You could use `name` as an extra key that's an actual key and have a `displayName` option for the cases where we don't have enough info.

Reply:
Not sure I understand the use case for that, and in any case it seems well beyond the scope of this PR so I think I'll revert to the "least disruptive" choice (the current state of this PR, but make `name` non-optional and default to `_unknown` for raw paths) and let others extend this later.